Update_demographics

rsmoke opened the pull request:
Manage the dempgrphics and updates the app to use the newest Stimulus

britaumich commented on app/admin/applicant_details.rb:
```diff
@@ -78,7 +79,11 @@
+        "#{d.demographic_name} - #{d.demographic_other}"
+      else
         d.demographic_name
+      end
```
The code in columns 81 - 86 is repeated three times.
If you replace demographic_name method in the Demographic model with 
```
  def demographic_name
    if self.demographic_other.present?
      Demographic.find(self.demographic).name + " - " + self.demographic_other
    else
      Demographic.find(self.demographic).name
    end
  end
```
in the app/admin/applicant_details.rb you can do
```
    column :demographic do |d| 
      d.demographic_name
    end
```

And line 155 will work correctly too.

rsmoke replied:
I updated methid names and made them more Rails idiomatic

britaumich commented on app/views/applicant_details/show.html.erb:
```diff
@@ -11,7 +11,7 @@
         <li><%= @us_citizen %></li>
-        <li>Demographic: <%= demographic_description(@applicant_detail) %></li>
+        <li>Demographic: <%= @applicant_detail.demographic_name %><%= " - #{@applicant_detail.demographic_other}" if @applicant_detail.demographic_other.present? && demographic_description(@applicant_detail) == "Other" %></li>
       </ul>
```
If demographic_name in the Demographic model is replaced with
```
def demographic_name
    if self.demographic_other.present?
      Demographic.find(self.demographic).name + " - " + self.demographic_other
    else
      Demographic.find(self.demographic).name
    end
  end
```
(see my comment for admin/demographic.rb)
then lines 14 can be replaced with:
```
<li>Demographic: <%= @applicant_detail.demographic_name %></li>
```
And in this case we don't need DemographicsHelper

britaumich reviewed (commented):
Beautiful app/javascript/controllers/demographic_other_controller.js script!

It would be nice to test the whole registration process together to see that all javascript files work correctly. (I have some messed up data for sessions and courses on my localhost - don't trust my tests).

Also the demographic reports should be updated to show demographic_other data.
For example, like this:
```
def complete_apps_demographic_report
      demographic_other_id = Demographic.find_by('name LIKE ?', "other").id.to_s
      query = "SELECT ad.country,
      (CASE WHEN ad.gender = '' THEN NULL ELSE 
      (SELECT genders.name FROM genders WHERE CAST(ad.gender AS UNSIGNED) = genders.id) END) as gender,
      e.year_in_school,
      (CASE ad.demographic WHEN '' THEN NULL 
            WHEN '" + demographic_other_id + "' THEN (SELECT CONCAT('Other - ', ad.demographic_other))
            ELSE 
            (SELECT demographics.name FROM demographics WHERE CAST(ad.demographic AS UNSIGNED) = demographics.id) END) AS demographic,
      e.international
      FROM enrollments AS e 
      LEFT JOIN users AS u ON e.user_id = u.id
      LEFT JOIN applicant_details AS ad ON ad.user_id = e.user_id
      WHERE e.application_status = 'application complete' AND e.campyear = #{CampConfiguration.active.last.camp_year} ORDER BY country, gender, year_in_school"
      title = "complete_apps_demographic_report"

      data = data_to_csv_demographic(query, title)
      respond_to do |format|
        format.html { send_data data, filename: "MMSS-report-#{title}-#{DateTime.now.strftime('%-d-%-m-%Y')}.csv"}
      end
 end

 def enrolled_student_demographic_report
      demographic_other_id = Demographic.find_by('name LIKE ?', "other").id.to_s
      query = "SELECT ad.country,
      (CASE WHEN ad.gender = '' THEN NULL ELSE 
      (SELECT genders.name FROM genders WHERE CAST(ad.gender AS UNSIGNED) = genders.id) END) as gender,
      e.year_in_school,
      (CASE ad.demographic WHEN '' THEN NULL 
            WHEN '" + demographic_other_id + "' THEN (SELECT CONCAT('Other - ', ad.demographic_other))
            ELSE 
            (SELECT demographics.name FROM demographics WHERE CAST(ad.demographic AS UNSIGNED) = demographics.id) END) AS demographic,
      e.international
      FROM enrollments AS e 
      LEFT JOIN users AS u ON e.user_id = u.id
      LEFT JOIN applicant_details AS ad ON ad.user_id = e.user_id
      WHERE e.application_status = 'enrolled' AND e.campyear = #{CampConfiguration.active.last.camp_year} ORDER BY country, gender, year_in_school"
      title = "enrolled_student_demographic_report"

      data = data_to_csv_demographic(query, title)
      respond_to do |format|
        format.html { send_data data, filename: "MMSS-report-#{title}-#{DateTime.now.strftime('%-d-%-m-%Y')}.csv"}
      end
 end
```

britaumich commented:
@rsmoke 
Note: I wish you did not put all refactoring (especially editing formating, quotes, and spaces) into one pull request. As you reminded us a hundred times before it complicates reviewing a lot. Of course, I understand why you did that.

Now a comment. In the [admin/](http://localhost:3000/admin/demographics) you display only Demographic.modifiable, and you changed 'Other' to 'protected'. That means that 'Other' is not displayed in the Demographic index view. 
If you don't display protected fields how to know all demographic records?
If I updated a field to be protected, I'll never see it again.

I tried to make a record protected and got an error: the field was updated but I guess the error was with a scope, because that record was not in Demographic.modifiable anymore.

Also, you added 'cannot contain punctuation' validation to the Demographic model. Now values like 
'Asian/Pacific Islander', 'Native American/American Indian', 'Black/African American' give a validation error when edited.

britaumich commented:
@rsmoke Demographic reports are not correct.
complete_apps_demographic_report doesn't show data from the 'demographic_other' field
<img width="664" alt="Screenshot 2024-11-11 at 7 22 17 PM" src="https://github.com/user-attachments/assets/3741afd7-0162-4b08-bcde-be6c6460840d">
enrolled_student_demographic_report displayed not correct data:

<img width="674" alt="Screenshot 2024-11-11 at 7 22 01 PM" src="https://github.com/user-attachments/assets/169dddce-afe2-42c3-b73c-22116d1e0f42">

rsmoke commented:
Fixed the reports and demographic other display. Fixed display of protected demographics. The dempgraphics that exist with punctuation will need to be updated

britaumich commented:
@rsmoke Reports works.

About protected demographic records. 
1. The update method to make a record protected in app/admin/demographics.rb (line 58) is not working - can't figure out why. 
2. But why do we need to make records other than 'Other" protected?
3. If we can make a record protected there is no way to make it unprotected and edit name and description. Why do we want that?
4. I think the 'Other' demographic should be protected (because we need it for demographic_other), but the rest of the records don't need to be protected and admins should not have the option to make them protected.

I tested demographic fields in the user's and admin's applicant details forms - all seem to be working.

I think we have to test the whole process in user and admin interfaces together because of all the refactoring and reformating you did in this pull request. It's not enough to just review the code, and I can trust myself doing all tests alone.

britaumich commented on app/views/course_preferences/_form.html.erb:
```diff
@@ -10,5 +10,4 @@
-<% end %>
-<%= javascript_pack_tag 'required_stars' %>
\ No newline at end of file
+<% end %>
```
There are several files in this pull request without an empty last line. Please see the red circle icon. If it doesn't bother you you can ignore my comment, but I like to get rid of these red icons in pull requests.
I think it is common practice in Rails and GitHub to end a file with a newline.

rsmoke commented:
Updated Demographics so Other is not editable and you cannot make a demographic protected from the Active Admin interface.

Added end lines to all the files that did not have them

made code less ocmplex in Reports file

britaumich commented on app/views/applicant_details/_form.html.erb:
```diff
@@ -69,15 +72,19 @@
+        flatpickr_min_date_value: (Date.today - 19.years).strftime('%Y-%m-%d'), 
+        flatpickr_max_date_value: (Date.today - 12.years).strftime('%Y-%m-%d') 
+      }, placeholder: "Select your birthdate" %>
     </div>
```
In the edit form the value of the birthdate field is not displayed.

rsmoke replied:
I am seeing it.
Try doing a bin/rails db:reset to be sure you don't have bad data

britaumich commented on app/javascript/controllers/recuploads_controller.js:
What is this controller? It says here that the file is empty, and I don't see it in the branch on my localhost.

britaumich reviewed (approved):
I approve this pull request, but we must test the whole process again.
I did test but I could easily miss some parts of the workflow.
